### ISSUE TITLE
Better Message for Corrupt Indicies

### ISSUE THREAD

**Issue by jeschu1** [microsoft/VFSForGit#1229]
A user received this error when they had a bad index.
 
Information {"Message":"System.ArgumentOutOfRangeException: Non-negative number required.\r\nParameter name: dstOffset\r\n   at System.Buffer.BlockCopy(Array src, Int32 srcOffset, Array dst, Int32 dstOffset, Int32 count)\r\n   at GVFS.Virtualization.Projection.GitIndexProjection.GitIndexParser.ReadPath(GitIndexEntry indexEntryData, Int32 replaceIndex, Int32 byteCount)\r\n   at GVFS.Virtualization.Projection.GitIndexProjection.GitIndexParser.ParseIndex(ITracer tracer, Stream indexStream, GitIndexEntry resuableParsedIndexEntry, Func`2 entryAction)\r\n   at GVFS.Virtualization.Projection.GitIndexProjection.GitIndexParser.RebuildProjection(ITracer tracer, Stream indexStream)\r\n   at GVFS.Virtualization.Projection.GitIndexProjection.BuildProjectionFromPath(ITracer tracer, String indexPath)\r\n   at GVFS.RepairJobs.RepairJob.TryParseIndex(String path, List`1 messages) in

Could it be more beneficial to run 'git ls-files' in and print out the error?
Can we handle this better?

@kewilford @wilbaker @jamill  for everyone's thoughts.

**Comment by jamill** [comment]
I think we have a copy of the corrupted index, so maybe we could try and understand what is wrong with it, and where VFS4G had trouble?

We should be able to do better than an `ArgumentOutOfRangeException` - even if it just indicating that we encountered a corrupt index and exiting the mount. If there was some guidance on what a user's next step might be to self mitigate, that could also be interesting... but I am not sure what the next step would nbe.

**Comment by kewillford** [comment]
I think this is a good place to try/catch and throw an application exception (new `GVFSException`?) with a message to run the `gvfs repair` which will fix the corrupt index.

**Comment by derrickstolee** [comment]
I think in this specific case, even `gvfs repair` failed with the same error. @jeschu1 do you remember? I think we ended up deleting the `.git/index` file an repairing after that.

**Comment by kewillford** [comment]
Yeah our parsing code right now doesn't handle corruption in the extensions.  I thought we had an issue for that but maybe that was just in my head.

**Comment by jeschu1** [comment]
@derrickstolee is correct we had to delete the index, we then ran repair i believe to recreate it.

If we fail to parse the index could we catch that and try to run 'git ls-files' showing the results?  I was thinking that would prove the index was corrupt and where.
